### ISSUE TITLE
`TCH004` false positive with submodule imports and rebinding

### ISSUE THREAD

**Issue by shiftinv** [charliermarsh/ruff#3821]
Importing the same top-level module through two submodule imports, one at runtime and one in a type-checking block, results in `TCH004` false positives.
This is certainly more of an edge case, seems like it could be related to #60 :thinking: 

Code:
```py
from __future__ import annotations

from typing import TYPE_CHECKING
import importlib.util

if TYPE_CHECKING:
    import importlib.machinery


def from_spec(spec: importlib.machinery.ModuleSpec):
    return importlib.util.module_from_spec(spec)
```

Running `ruff check --isolated --select TCH test.py` (v0.0.260) produces the following error:
```py
test.py:7:12: TCH004 Move import `importlib.machinery` out of type-checking block. Import is used for more than type hinting.
```

I don't know a whole lot about ruff's internals, but my assumption is that the `importlib.machinery` import in the type-checking block overwrites the `importlib` binding, and the `importlib.util.module_from_spec` call later resolves to that.

**Comment by charliermarsh** [comment]
Hmm, what's happening here is that both submodule imports resolve to `importlib`, since submodule imports can resolve to an import of the parent -- e.g., this works:

```py
>>> import urllib.request
>>> urllib.response
<module 'urllib.response' from '/usr/local/Cellar/python@3.11/3.11.0/Frameworks/Python.framework/Versions/3.11/lib/python3.11/urllib/response.py'>
```

(I'm probably getting a bunch of terminology wrong, but in short, like Pyflakes, we tend to treat `import importlib.util` as identical to `import importlib`, which seems to be incorrect here.)

**Comment by charliermarsh** [comment]
I'm not 100% sure how to fix in a general way but I agree that it's a bug here.

**Comment by shiftinv** [comment]
Yeah, it's a bit of a weird bug. Treating `import a.b.c` like `import a` is completely fair for a linter, imo - with possible import side effects it's probably better to be  lenient with these things.

The original issue can be solved easily by just moving the `importlib.machinery` import, though I suppose at that point one could argue that it should emit a `TCH003` (`Move standard library import {} into a type-checking block`).

In the end it depends on how "using an import" is defined (e.g. does `a.b.func()` use `a.b` or `a`?) - at least in the general case, the current behavior is a good compromise, now that I think about it.
The alternative, which is being more strict about submodule imports, might end up creating a lot more different false positives, so... not sure.

**Comment by Avasam** [comment]
Edit: Currently on Ruff 0.14.4 and I am no longer getting a false-positive with the example below

--- 

Similar issue here with Python 3.9 and opencv-python, which just added a `typing` module (not released to pypi yet). But since it's not picked up by PyInstaller (yet), I wrapped the import in a `TYPE_CHECKING` check:

```py
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    import cv2.typing

def is_blank(image: cv2.typing.MatLike):
    # do stuff here
    pass
```

Note that at runtime (normal python runtime outside of PyInstaller), the typing module exists, but is empty (https://github.com/opencv/opencv/issues/23782). So I know that this import is only used as type-hint otherwise `cv2.typing.MatLike` would fail.

**Comment by dimbleby** [comment]
Another example that shows up pretty often is with types from `collections.abc` eg:
```python
  from typing import TYPE_CHECKING

  if TYPE_CHECKING:
     from collections.abc import Iterable

  def fun() -> Iterable[int]:
      yield from range(5)
```
```
foo.py:6:33: TCH004 [*] Move import `collections.abc.Iterable` out of type-checking block. Import is used for more than type hinting.`
```
which is clearly not right

**Comment by charliermarsh** [comment]
@dimbleby - That actually _is_ correct -- if you try running that code, you will get:

```
Traceback (most recent call last):
  File "/Users/crmarsh/workspace/ruff/foo.py", line 6, in <module>
    def fun() -> Iterable[int]:
                 ^^^^^^^^
NameError: name 'Iterable' is not defined
```

Python evaluates types in function annotations at runtime as they're added to `__annotations__`. If you add `from __future__ import annotations`, Ruff will do the right thing. In a future version, we may add the ability to automatically quote those annotations (`"Iterable[int]"`) as part of the rule.

**Comment by scur-iolus** [comment]
I've got similar issues in a Django project, but it seems to be even worse, since I'm not even importing parent modules. Should I report another issue?

Below is an example with a class of [django-simple-history](https://django-simple-history.readthedocs.io/en/latest/). This being said, the same false positive error appears with some of my own classes:
```python
from typing import TYPE_CHECKING

from django.db import models

if TYPE_CHECKING:
    from simple_history.models import HistoricalChanges # TCH004 Move import `simple_history.models.HistoricalChanges` out of type-checking block [...]


class EntryManager(models.Manager):
    def my_function(self, previous: HistoricalChanges | None) -> int:
        ...

# that's all in this file!
```

![Screenshot at 09-24-08](https://github.com/astral-sh/ruff/assets/57783432/8dbabe9f-4906-47c3-bc05-9daa88af41c2)


I'm using `ruff 0.1.0`, `Python 3.11`. I don't think I had encountered the issue before upgrading to `0.1`, but I might be wrong.
